### PR TITLE
Add evmone t8n tool

### PR DESCRIPTION
**Pull request by chfast**
The t8n is "state transition" CLI interface used to check and generate
JSON tests. It reads and outputs specified JSON files.
See https://ethereum-tests.readthedocs.io/en/develop/t8ntool-ref.html.

### CODE REVIEW

**Comment by chfast** [comment]
First example to run:

- geth
  ```
  retesteth -t 'GeneralStateTests/stExample' -- --clients t8ntool --testpath ~/Projects/ethereum/tests --singletest add11 --singlenet Berlin --verbosity 6
  ```

- evmone
  ```
  retesteth -t 'GeneralStateTests/stExample' -- --datadir ~/Projects/ethereum/evmone/config/retesteth --clients evmone --testpath ~/Projects/ethereum/tests --singletest add11 --singlenet Berlin --verbosity 6
  ```

**Comment by chfast** [comment]
The example isn't working currently because evmone does not produce expected transaction receipts (empty array in the result). The receipt is match with a transaction by transaction hash which evmone is not able to compute yet.

File: test/statetest/secret_sender_map.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+#include "secret_sender_map.hpp"
+#include <cassert>
+#include <map>
+
+namespace evmone::test
+{
+using namespace evmc;
+
+evmc::address sender_address(const bytes32& secret_key)
```
Comment (axic) [inline test/statetest/secret_sender_map.cpp:13]: Is this `private_key` -> `address` lookup table?

File: test/statetest/statetest.hpp [hunk 2]
```diff
@@ -59,8 +63,59 @@ struct StateTransitionTest
...
+}
+
+inline size_t countl_zero_bytes(uint64_t x)
+{
+    auto upper = uint32_t(x >> 32);
+    auto lower = uint32_t(x & 0xFFFFFFFF);
+    return upper ? countl_zero_bytes(upper) : 4 + countl_zero_bytes(lower);
+}
+
+inline std::string hex(uint64_t v)
```
Comment (axic) [inline test/statetest/statetest.hpp:105]: I'd call this something like `shortest_hex` or something similar to signal intent.

File: test/statetest/statetest_loader.cpp [hunk 3]
```diff
@@ -11,11 +12,6 @@ namespace evmone::test
 namespace json = nlohmann;
 using evmc::from_hex;
 
-namespace
```
Comment (axic) [inline test/statetest/statetest_loader.cpp:14]: Shouldn't these stay in the anonymous namespace?

File: test/t8n/t8n.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <string_view>
+
+namespace fs = std::filesystem;
+namespace json = nlohmann;
+using namespace evmone;
+
+namespace evmone::state
```
Comment (axic) [inline test/t8n/t8n.cpp:15]: Do we need this available there or can this be anonymous?

File: test/t8n/t8n.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            }
+
+            j_result["logsHash"] = "0x" + hex(keccak256(rlp::encode(txs_logs)));
+            j_result["stateRoot"] = "0x" + hex(state::mpt_hash(state.get_accounts()));
+        }
+
+        j_result["logsBloom"] = "0x" + std::string(512, '0');
+        j_result["receiptsRoot"] =
+            "0x0000000000000000000000000000000000000000000000000000000000000000";
+        j_result["txRoot"] = "0x0000000000000000000000000000000000000000000000000000000000000000";
```
Comment (axic) [inline test/t8n/t8n.cpp:154]: Could we have an overloaded version of `hex` to support various types, and then use something like `hex(intx::uint256{0})`?

Comment (axic) [reply test/t8n/t8n.cpp:154]: Or perhaps have these constants pre-defined at the top of the file and reused multiple times (e.g. `NULL_HEXSTRING_160`, `NULL_HEXSTRING_256`, `NULL_HEXSTRING_512`)

File: test/statetest/statetest_loader.cpp (hunk 3 above)
Comment (rodiazet) [reply test/statetest/statetest_loader.cpp:14]: No, because `from_json` is now also defined in header.

File: test/t8n/t8n.cpp (hunk 4 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:15]: Unfortunately it has to be there. Similar to `statetest_runner.cpp`. In other case template functions encode instantiation fails.

File: test/t8n/t8n.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+        std::ofstream{output_dir / output_result_file} << j_result;
+        std::cerr << "---\nRESULT:\n" << std::setw(2) << j_result << "\n";
+
+        json::json j_alloc;
+        for (const auto& [addr, acc] : state.get_accounts())
+        {
+            j_alloc["0x" + hex(addr)]["nonce"] = "0x" + test::utils::hex(acc.nonce);
```
[the remark this answers is not in the document]
Comment (rodiazet) [reply test/t8n/t8n.cpp:186]: I introduced changes to make it less confusing.

File: test/t8n/t8n.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                for (auto& j_tx : j_txs)
+                {
+                    state::Transaction tx = test::from_json<state::Transaction>(j_tx);
+                    auto const [tx_logs, gas_used] = state::transition(state, block, tx, rev, vm);
+                    if (tx_logs.has_value())
+                    {
+                        txs_logs.insert(txs_logs.begin(), (*tx_logs).begin(), (*tx_logs).end());
+                        auto& j_receipt = j_result["receipts"][j_result["receipts"].size()];
+                        j_receipt["transactionHash"] = j_tx["hash"];
+                        j_receipt["gasUsed"] = tu::shortest_hex(uint64_t(gas_used));
```
Comment (axic) [inline test/t8n/t8n.cpp:117]: What if `int64_t gas_used` is negative (due to a failure)?

File: test/t8n/t8n.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                        j_receipt["cumulativeGasUsed"] = j_receipt["gasUsed"];
+
+                        j_receipt["blockHash"] = NULL_HEXSTRING_32;
+                        j_receipt["contractAddress"] = NULL_HEXSTRING_20;
+                        j_receipt["logsBloom"] = NULL_HEXSTRING_256;
+                        j_receipt["logs"] = json::json::array();  // FIXME: Add to_json<state:Log>
+                        j_receipt["root"] = "";
```
Comment (axic) [inline test/t8n/t8n.cpp:124]: Is this really expected to be empty?

File: test/t8n/t8n.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        // Print out current state to outAlloc file
+        json::json j_alloc;
+        for (const auto& [addr, acc] : state.get_accounts())
+        {
+            j_alloc[tu::hex(addr)]["nonce"] = test::utils::shortest_hex(acc.nonce);
+            for (const auto& [key, val] : acc.storage)
+                if (!is_zero(val.current))
+                    j_alloc[tu::hex(addr)]["storage"][tu::hex(key)] = tu::hex(val.current);
+
+            j_alloc[tu::hex(addr)]["code"] = tu::hex({acc.code.data(), acc.code.size()});
```
Comment (axic) [inline test/t8n/t8n.cpp:160]: This has no `0x` prefix?

File: test/t8n/t8n.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+namespace evmone::state
+{
+/// Defines how to RLP-encode a Log. This is only needed to compute "logs hash".
+inline bytes rlp_encode(const Log& log)
+{
+    return rlp::encode_tuple(log.addr, log.topics, log.data);
+}
+}  // namespace evmone::state
+
+int main(int argc, const char* argv[])
```
Comment (axic) [inline test/t8n/t8n.cpp:29]: We should have a unit test for this (mostly for checking the input -> output mapping is unchanged). Question is: is it better to use a CLI test or make this into an internal library first?

File: test/t8n/t8n.cpp (hunk 8 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:124]: It's a minimal working version. A lot of fields are required by retesteth but never used.

File: test/t8n/t8n.cpp (hunk 10 above)
Comment (rodiazet) [reply test/t8n/t8n.cpp:29]: I would prefer the second. @chfast What do you think?

File: test/t8n/t8n.cpp (hunk 9 above)
Comment (axic) [reply test/t8n/t8n.cpp:160]: One more idea: we could have an overloaded helper `hex` defined in this file at the top for each type we care of, so that the body here can just use `hex()` solely. This way we can inspect what types are used and how they get translated.

Comment (rodiazet) [reply test/t8n/t8n.cpp:160]: But if these helper needs to use hex impl from different libs like evmc it will be a mess with specifying which one we want to use.

Comment (axic) [reply test/t8n/t8n.cpp:160]: I meant something like this:
```cpp
namespace {
inline bytes hex(string_view s) {
    return evmc::hex(s);
}
inline bytes hex(intx::u256 v) {
    return intx::hex(v);
}
}
```

Comment (rodiazet) [reply test/t8n/t8n.cpp:160]: Sure I understand but we will have to remove a couple of `using namespace evmc` etc.

File: test/statetest/statetest.hpp [hunk 11]
```diff
@@ -59,8 +63,66 @@ struct StateTransitionTest
...
+    const auto prefix = std::string(add_0x ? "0x" : "");
+    size_t num_lz_bytes = countl_zero_bytes(v);
+
+    if (num_lz_bytes == 8)
+        return prefix + "0";
+
+    std::string str;
+
+    for (auto i = num_lz_bytes; i < 8; ++i)
+        str += evmc::hex(static_cast<uint8_t>(v >> 8 * (8 - i - 1)));
```
Comment (axic) [inline test/statetest/statetest.hpp:116]: Could also just translate it to string first and then doing a substring to the first non-`0` character.

File: test/statetest/statetest.hpp [hunk 12]
```diff
@@ -59,8 +63,66 @@ struct StateTransitionTest
...
+    auto lower = uint32_t(x & 0xFFFFFFFF);
+    return upper ? countl_zero_bytes(upper) : 4 + countl_zero_bytes(lower);
+}
+
+inline std::string shortest_hex(uint64_t v, bool add_0x = true)
+{
+    const auto prefix = std::string(add_0x ? "0x" : "");
+    size_t num_lz_bytes = countl_zero_bytes(v);
+
+    if (num_lz_bytes == 8)
```
Comment (axic) [inline test/statetest/statetest.hpp:110]: This could be done with `if (v == 0)`.

File: test/statetest/statetest.hpp (hunk 11 above)
Comment (axic) [reply test/statetest/statetest.hpp:116]:
```cpp
inline std::string shortest_hex(uint64_t v, bool add_prefix = true)
{
    if (v == 0) {
        return add_prefix ? "0x0" : "0";
    }

    auto value = hex(v, false);
    // String leading zeroes.
    if (size_t pos = value.find_first_not_of("0"); pos != string::npos) {
        value = value.substr(pos);
    }
    return (add_prefix ? "0x" : "") + value;
}
```

File: test/statetest/statetest_loader.cpp [hunk 13]
```diff
@@ -69,19 +65,89 @@ state::AccessList from_json<state::AccessList>(const json::json& j)
     return o;
 }
 
+inline uint64_t calculate_current_base_fee_eip1559(
+    uint64_t parent_gas_used, uint64_t parent_gas_limit, uint64_t parent_base_fee)
+{
+    static const auto BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;
+    static const auto ELASTICITY_MULTIPLIER = 2;
+
+    uint64_t base_fee = 0;
```
Comment (axic) [inline test/statetest/statetest_loader.cpp:74]: Is 64-bit precision enough for these? Someone should double check EIP-1559.

**Comment by axic** [comment]
Should remove `.config.swp` before merging. Also why is there a need to duplicate the config files, aren't they part of reteseth?

File: test/statetest/statetest_loader.cpp (hunk 13 above)
Comment (rodiazet) [reply test/statetest/statetest_loader.cpp:75]: TODO added.

**Comment by rodiazet** [comment]
> Should remove `.config.swp` before merging. Also why is there a need to duplicate the config files, aren't they part of reteseth?

Definitely configs should be moved to retesteth.

File: test/statetest/statetest.hpp [hunk 14]
```diff
@@ -59,8 +63,54 @@ struct StateTransitionTest
...
+inline std::string shortest_hex(uint64_t v, bool add_prefix = true)
+{
+    if (v == 0)
+    {
+        return add_prefix ? "0x0" : "0";
+    }
+
+    std::string value;
+    for (auto i = 0; i < 8; ++i)
+        value += evmc::hex(static_cast<uint8_t>(v >> 8 * (8 - i - 1)));
```
Comment (axic) [inline test/statetest/statetest.hpp:94]: Hm, can't this actually be done with?
```suggestion
    value = intx::hex(intx::uint256{v});
```

Comment (rodiazet) [reply test/statetest/statetest.hpp]: Done

File: test/t8n/t8n.cpp [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+#include "../statetest/statetest.hpp"
+#include <evmone/evmone.h>
+#include <nlohmann/json.hpp>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <string_view>
+
+namespace fs = std::filesystem;
+namespace json = nlohmann;
```
Comment (axic) [inline test/t8n/t8n.cpp:12]: Why not?
```suggestion
namespace json = nlohmann::json;
```

Would remove the `json::json::...` cases.

Comment (axic) [reply test/t8n/t8n.cpp:12]: Ah ok, this is consistent with `statetest_loader`, can change later then.

File: test/t8n/t8n.cpp [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+            {
+                j_result["receipts"] = json::json::array();
+                j_result["rejected"] = json::json::array();
+                auto idx = 0;
+                for (auto& j_tx : j_txs)
+                {
+                    state::Transaction tx = test::from_json<state::Transaction>(j_tx);
+                    auto const result = state::transition(state, block, tx, rev, vm);
+                    const auto receipt = get_if<state::TransactionReceipt>(&result);
+                    if (receipt != nullptr)
```
Comment (axic) [inline test/t8n/t8n.cpp:113]: I'd swap and have the failure case first, as that is shorter.

File: test/statetest/statetest.hpp [hunk 17]
```diff
@@ -59,8 +63,52 @@ struct StateTransitionTest
...
+state::Transaction from_json<state::Transaction>(const json::json& j);
+
 StateTransitionTest load_state_test(const fs::path& test_file);
 
 void run_state_test(const StateTransitionTest& test, evmc::VM& vm);
 
+namespace utils
+{
+
+inline std::string shortest_hex(uint64_t v, bool add_prefix = true)
```
Comment (chfast) [inline test/statetest/statetest.hpp:85]: Is this really needed? I believe `intx::hex` is already producing the "shorted" hex as expected for any integer.

File: test/statetest/statetest.hpp [hunk 18]
```diff
@@ -59,8 +63,52 @@ struct StateTransitionTest
...
+    }
+    return (add_prefix ? "0x" : "") + value;
+}
+
+inline std::string hex(bytes_view bs, bool add_prefix = true)
+{
+    return (add_prefix ? "0x" : "") + evmc::hex(bs);
+}
+
+inline std::string hex(const intx::uint<256>& v, bool add_prefix = true)
```
Comment (chfast) [inline test/statetest/statetest.hpp:107]: It looks we are just wrapping existing `hex` functions to add a prefix. How about a template:
```cpp
template<typename T>
auto hex0x(const T& v)
{
    return "0x" + hex(v);
}
```

Comment (rodiazet) [reply test/statetest/statetest.hpp]: done

File: test/statetest/statetest.hpp (hunk 17 above)
Comment (rodiazet) [reply test/statetest/statetest.hpp]: you are right. Fixed.

Comment (axic) [reply test/statetest/statetest.hpp]: In fairness, we slowly arrived at the point of using `intx::hex()` which then allowed this further simplification 😅

File: lib/evmone/baseline.cpp [hunk 19]
```diff
@@ -356,6 +356,11 @@ evmc_result execute(const VM& vm, ExecutionState& state, const CodeAnalysis& ana
 evmc_result execute(evmc_vm* c_vm, const evmc_host_interface* host, evmc_host_context* ctx,
     evmc_revision rev, const evmc_message* msg, const uint8_t* code, size_t code_size) noexcept
 {
+    if (rev >= EVMC_SHANGHAI && is_eof_code({code, code_size}) &&
```
Comment (chfast) [inline lib/evmone/baseline.cpp:359]: Why is this needed?

File: lib/evmone/errors.hpp [hunk 20]
```diff
@@ -0,0 +1,80 @@
+// evmone: Fast Ethereum Virtual Machine implementation
```
Comment (chfast) [inline lib/evmone/errors.hpp:1]: This file should go to test/state.

File: test/state/state.cpp [hunk 21]
```diff
@@ -53,42 +54,48 @@ int64_t compute_tx_intrinsic_cost(evmc_revision rev, const Transaction& tx) noex
...
-        return -1;  // Origin must not be a contract (EIP-3607).
+        return make_error_code(SENDER_NOT_EOA);  // Origin must not be a contract (EIP-3607).
 
     if (sender_acc.nonce == Account::NonceMax)
-        return -1;
+        return make_error_code(NONCE_HAS_MAX_VALUE);
 
     if (rev >= EVMC_SHANGHAI && !tx.to.has_value() && tx.data.size() > max_initcode_size)
-        return -1;  // initcode size is limited by EIP-3860.
+        return make_error_code(INIT_CODE_SIZE_LIMIT_EXCEEDED);  // initcode size is limited by
```
Comment (chfast) [inline test/state/state.cpp:82]: Move comment before `if`.

File: test/t8n/t8n.cpp [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+        {
+            const auto j = json::json::parse(std::ifstream{env_file});
+            block = test::from_json<state::BlockInfo>(j);
+        }
+
+        auto state = pre_state;
+
+        json::json j_result;
+        // FIXME: Calculate difficulty properly
+        j_result["currentDifficulty"] = "0x20000";
```
Comment (chfast) [inline test/t8n/t8n.cpp:86]: Is this right?

File: config/retesteth/evmone/config [hunk 23]
```diff
@@ -0,0 +1,239 @@
+{
+    "name" : "evmone",
+    "socketType" : "tranition-tool",
+    "socketAddress" : "start.sh",
+    "transactionsAsJson" : true,
+    "checkLogsHash" : false,
```
Comment (chfast) [inline config/retesteth/evmone/config:6]: Do you want to try this?

File: lib/evmone/baseline.cpp (hunk 19 above)
Comment (rodiazet) [reply lib/evmone/baseline.cpp:359]: For code equals `0xfe00` (EOF) function `read_valid_eof1_header` (called by `analyze`) assumes that there is something else but it's the whole code so it tries to read data from the outside of the code section.

File: test/t8n/t8n.cpp (hunk 15 above)
Comment (axic) [reply test/t8n/t8n.cpp:12]: Right, can we just include the namespace (`using namespace nlohmann`)?

File: test/t8n/t8n.cpp [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+            evmc::VM vm{evmc_create_evmone(), {{"O", "0"}}};
+
+            std::vector<state::Log> txs_logs;
+
+            if (j_txs.is_array())
+            {
+                j_result["receipts"] = json::json::array();
+                j_result["rejected"] = json::json::array();
+                auto idx = 0;
+                for (auto& j_tx : j_txs)
```
Comment (axic) [inline test/t8n/t8n.cpp:101]: Is this modifying `j_txs` or can this be const?

**Review by rodiazet — APPROVED** [review]

File: test/statetest/statetest_loader.cpp [hunk 25]
```diff
@@ -255,6 +255,38 @@ static void from_json(const json::json& j, StateTransitionTest& o)
...
+    state::Transaction o;
+    o.data = from_json<bytes>(j.at("input"));
+    o.gas_limit = from_json<int64_t>(j.at("gas"));
+    o.value = from_json<intx::uint256>(j.at("value"));
+
+    if (!j.at("to").get<std::string>().empty())
+        o.to = from_json<evmc::address>(j.at("to"));
+
+    o.sender = from_json<evmc::address>(j.at("sender"));
+    // FIXME: Get max gas price
```
Comment (axic) [inline test/statetest/statetest_loader.cpp:270]: Ia this commit still valid?

Comment (chfast) [reply test/statetest/statetest_loader.cpp]: No.
